book/src/dev/rfcs/0012-value-pools.md

```diff
@@ -0,0 +1,497 @@
+pub fn value_balance(&self, utxos: &HashMap<transparent::OutPoint, Utxo>) -> Result<ValueBalance<NegativeAllowed>, Err> {
+    ..
+    // As soon as we have the value balance for all the pools
+    // check the remaining transaction value consensus rule:
+    value_balance.remaining_transaction_value()?;
+
+    Ok(value_balance);
+}
+```
+
+### Create a method in `Block` that returns `ValueBalance<NegativeAllowed>` for the block
+
+- Method location is at `zebra-chain/src/block.rs`.
+- Method will make use of `Transaction::value_balance` method created before.
+
+```rust
+/// utxos must contain the utxos of every input in the transaction,
+/// including UTXOs created by a transaction in this block,
+/// then spent by a later transaction that's also in this block.
+pub fn value_balance(&self, utxos: &HashMap<transparent::OutPoint, Utxo>) -> ValueBalance<NegativeAllowed> {
+    self.transactions()
+        .map(Transaction::value_balance)
+        .sum()
+        .expect("Each block should have at least one coinbase transaction")
+}
+```
+
+### Pass the value balance for this block from the consensus into the state
+
+- Add a new field into `PreparedBlock` located at `zebra-state/src/request.rs`, this is the `NonFinalized` section of the state.
+
+```rust
+pub struct PreparedBlock {
+    ..
+    /// The value balances for each pool for this block.
+    pub block_value_balance: ValuePool<NegativeAllowed>,
+}
+```
+- In `zebra-consensus/src/block.rs` pass the value balance to the zebra-state:
+
+```rust
+let block_value_balance = block.value_balance();
+let prepared_block = zs::PreparedBlock {
+    ..
+    block_value_balance,
+};
+```
+
+### Add a value pool into the state `Chain` struct
+
+- This is the value pool for the non finalized part of the blockchain.
+- Location of the `Chain` structure where the pool field will be added: `zebra-state/src/service/non_finalized_state/chain.rs`
+
+```rust
+pub struct Chain {
+    ..
+    /// The chain value pool balance at the tip of this chain.
+    value_pool: ValueBalance<NonNegative>,
+}
+```
+- Add a new argument `finalized_tip_value_balance` to the `commit_new_chain()` method located in the same file.
+- Pass the new argument to the Chain in:
+
+```rust
+let mut chain = Chain::new(finalized_tip_history_tree, finalized_tip_value_balance);
+```
+
+Note: We don't need to pass the finalized tip value balance into the `commit_block()` method.
+
+### Check the consensus rules when the chain is updated or reversed
+
+- Location: `zebra-state/src/service/non_finalized_state/chain.rs`
+
+```rust
+impl UpdateWith<PreparedBlock> for ValueBalance<NegativeAllowed> {
+    fn update_chain_state_with(&mut self, value_balance: &ValueBalance<NegativeAllowed>) -> Result<(), Err> {
+        self += value_balance?;
+    }
+
+    fn revert_chain_state_with(&mut self, value_balance: &ValueBalance<NegativeAllowed>) -> Result<(), Err> {
+        self -= value_balance?;
+    }
+}
```

Comment:
The overall idea here is solid, we just need to tweak a few details.

The `UpdateWith` trait is a bit tricky, it's always implemented for `Chain`, but the type we're updating changes.

Also the addition and subtraction of `ValueBalance` won't quite work, because `AddAssign` and `SubAssign` are only defined for `Result<Amount<C>, _>`.

```suggestion
impl UpdateWith<ValueBalance<NegativeAllowed>> for Chain {
    fn update_chain_state_with(&mut self, value_balance: &ValueBalance<NegativeAllowed>) -> Result<(), Err> {
        self.value_pool = (self.value_pool + value_balance)?;
        Ok(())
    }

    fn revert_chain_state_with(&mut self, value_balance: &ValueBalance<NegativeAllowed>) -> Result<(), Err> {
        self.value_pool = (self.value_pool + value_balance)?;
        Ok(())
    }
}
```

Reply:
https://github.com/ZcashFoundation/zebra/pull/2430/commits/f05bde4afca6231782c78d1fee8422cc798e9099